src/Guard.php

```diff
@@ -215,131 +205,126 @@ public function generateToken()
+            $token = $this->storage[$name];
 
-        $request = $this->attachRequestAttributes($request, $pair);
+            if (function_exists('hash_equals')) {
+                $valid = hash_equals($token, $value);
+            } else {
+                $valid = $token === $value;
+            }
+        }
 
-        return $request;
+        return $valid;
     }
 
     /**
-     * Validate CSRF token from current request
-     * against token value stored in $_SESSION
-     *
-     * @param  string $name  CSRF name
-     * @param  string $value CSRF token value
-     *
-     * @return bool
+     * @return string
      */
-    public function validateToken($name, $value)
+    public function getTokenName(): ?string
     {
-        $token = $this->getFromStorage($name);
-        if (function_exists('hash_equals')) {
-            $result = ($token !== false && hash_equals($token, $value));
-        } else {
-            $result = ($token !== false && $token === $value);
-        }
-
-        // If we're not in persistent token mode, delete the token.
-        if (!$this->persistentTokenMode) {
-            $this->removeFromStorage($name);
-        }
+        return $this->keyPair[$this->getTokenNameKey()] ?? null;
+    }
 
-        return $result;
+    /**
+     * @return string
+     */
+    public function getTokenValue(): ?string
+    {
+        return $this->keyPair[$this->getTokenValueKey()] ?? null;
     }
 
     /**
-     * Create CSRF token value
-     *
      * @return string
      */
-    protected function createToken()
+    public function getTokenNameKey(): string
     {
-        return bin2hex(random_bytes($this->strength));
+        return $this->prefix . '_name';
     }
 
     /**
-     * Save token to storage
-     *
-     * @param  string $name  CSRF token name
-     * @param  string $value CSRF token value
+     * @return string
      */
-    protected function saveToStorage($name, $value)
+    public function getTokenValueKey(): string
     {
-        $this->storage[$name] = $value;
+        return $this->prefix . '_value';
     }
 
     /**
-     * Get token from storage
-     *
-     * @param  string      $name CSRF token name
-     *
-     * @return string|bool CSRF token value or `false` if not present
+     * @return bool
      */
-    protected function getFromStorage($name)
+    public function getPersistentTokenMode(): bool
     {
-        return isset($this->storage[$name]) ? $this->storage[$name] : false;
+        return $this->persistentTokenMode;
     }
 
     /**
-     * Get the most recent key pair from storage.
-     *
-     * @return string[]|null Array containing name and value if found, null otherwise
-     */    
-    protected function getLastKeyPair()
+     * @return string[]|null
+     */
+    protected function getLastKeyPair(): ?array
     {
         // Use count, since empty ArrayAccess objects can still return false for `empty`
         if (count($this->storage) < 1) {
```

Comment:
Not sure if that's an issue: The function `count` expects `array|Countable`, but we give `array|ArrayAccess`. How about
```php
        if ((is_countable($this->storage) && count($this->storage) < 1) || empty($this->storage)) {
            return null;
        }
```

Reply:
Ah no, `is_countable` is PHP 7.3. We need another solution.

Reply:
Probably:
```php
        if (is_array($this->storage) && empty($this->storage)) {
            return null;
        }
        if ($this->storage instanceof ArrayAccess) {
            if (!($this->storage instanceof Countable)) {
                throw new RuntimeException('The Csrf storage is not countable.');
            }
            if (count($this->storage) < 1) {
                return null;
            }
        }
```